### DISCUSSION TITLE
Tests with `WebApplicationFactory` do not send OpenTelemetry data

### DISCUSSION THREAD

**Discussion by fdcastel** [xunit/xunit#3045] [Question] [answered]
Does anyone else experience a similar issue? 

When running tests with `WebApplicationFactory` in a system integrated with OpenTelemetry, in many cases the telemetry data is not sent (often seen in shorter, quicker tests). 

I did notice that in slower tests, the telemetry data gets sent. 

It seems that the SUT process is killed before it have a chance to send the telemetry data. There is a way to wait for this?

I am seeking guidance on how to conduct a more thorough investigation into this matter.

**Comment by bradwilson** [comment, marked as answer]
> It seems that the SUT process is killed before it have a chance to send the telemetry data. There is a way to wait for this?

We don't have any support (explicit or implicit) for `WebApplicationFactory` or OpenTelemetry. By the way you're describing it, it does sound like OpenTelemetry is typically sending messages in the background on a timer or some sort, and the shutdown of `WebApplicationFactory` (or similar) is not properly notifying OpenTelemetry that it should clean itself up (aka, send the remaining telemetry messages).

I would suggest opening an issue against `WebApplicationFactory` to see if they have suggestions on whether it's a bug on their side that needs to be fixed, or perhaps there's a cleanup operation you need to trigger that is missing.

Good luck!
